org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/IndexManager.java

```diff
@@ -1155,26 +1161,11 @@ public void saveIndex(Index index) throws IOException {
 			return;
 		}
 	}
+	IPath containerPath = new Path(index.containerPath);
+	if (hasPendingJobMatching(job -> (job instanceof IndexRequest) && (((IndexRequest) job).containerPath.equals(containerPath)))) {
+		return;
+	}
 	synchronized (this) {
```

Comment:
The scope of the `synchronized(this)` is reduced and allows non-sync changes between the check / return and following code. I will move the check above into the sync scope back.